File: testdata/fixtures/php/vulnerable/ssrf.php
<?php
// GTSS-PHP-002: SSRF via file_get_contents and curl

function fetch_url($url) {
    // Vulnerable: user-controlled URL
    $content = file_get_contents($url);
    return $content;
}

function proxy_request() {
    $target = $_GET['url'];
    // Vulnerable: SSRF via fopen
    $handle = fopen($target, 'r');
    $data = fread($handle, 8192);
    fclose($handle);
    echo $data;
}

function curl_fetch($userUrl) {
    $ch = curl_init();
    // Vulnerable: user-controlled CURLOPT_URL
    curl_setopt($ch, CURLOPT_URL, $userUrl);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    $result = curl_exec($ch);
    curl_close($ch);
    return $result;
}

File: batou-rules/testdata/fixtures/zig/vulnerable/ssrf.zig
const std = @import("std");

// SSRF via std.http.Client.fetch with user-controlled URL
pub fn proxyRequest(allocator: std.mem.Allocator, user_url: []const u8) !void {
    var client = std.http.Client{ .allocator = allocator };
    defer client.deinit();

    const result = try client.fetch(.{
        .url = user_url,
    });
    _ = result;
}

// SSRF via std.http.Client.open with tainted URI
pub fn openConnection(allocator: std.mem.Allocator, target: []const u8) !void {
    var client = std.http.Client{ .allocator = allocator };
    defer client.deinit();

    const uri = try std.Uri.parse(target);
    var req = try client.open(.GET, uri, .{
        .allocator = allocator,
    });
    defer req.deinit();
}
